robot-server/robot_server/protocols/protocol_store.py

```diff
@@ -104,45 +240,131 @@ def remove(self, protocol_id: str) -> ProtocolResource:
+    and keep them in memory.
+
+    Params:
+        expected_protocol_ids: The ID of every protocol for which to compute a
+            `ProtocolSource`.
+        protocols_directory: A directory containing one subdirectory per protocol
+            named by protocol ID. Scanned for files to pass to `protocol_reader`.
+        protocol_reader: An interface to use to compute `ProtocolSource`s.
+
+    Returns:
+        A map from protocol ID to computed `ProtocolSource`.
+
+    Raises:
+        Exception: This is not expected to raise anything,
+            but it might if a software update makes ProtocolReader reject files
+            that it formerly accepted.
+    """
+    sources_by_id: Dict[str, ProtocolSource] = {}
+
+    directory_members = [m async for m in protocols_directory.iterdir()]
+    directory_member_names = set(m.name for m in directory_members)
+    extra_members = directory_member_names - expected_protocol_ids
+    missing_members = expected_protocol_ids - directory_member_names
+
+    if extra_members:
+        # Extra members may be left over from prior interrupted writes
+        # and other kinds of failed insertions.
+        _log.warning(
+            f"Unexpected files or directories inside protocol storage directory:"
+            f" {extra_members}."
+            f" Ignoring them."
+        )
 
-        return deleted_resource
+    if missing_members:
+        raise SubdirectoryMissingError(
+            f"Missing subdirectories for protocols: {missing_members}"
+        )
+
+    async def compute_source(
+        protocol_id: str, protocol_subdirectory: AsyncPath
+    ) -> None:
+        # Given that the expected protocol subdirectory exists,
+        # we trust that the files in it are correct.
+        # No extra files, and no files missing.
+        #
+        # This is a safe assumption as long as:
+        #  * Nobody has tampered with file the storage.
+        #  * We don't try to compute the source of any protocol whose insertion
+        #    failed halfway through and left files behind.
+        protocol_files = [Path(f) async for f in protocol_subdirectory.iterdir()]
+        protocol_source = await protocol_reader.read_saved(
+            files=protocol_files, directory=Path(protocol_subdirectory)
+        )
+        sources_by_id[protocol_id] = protocol_source
 
+    async with create_task_group() as task_group:
+        # Use a TaskGroup instead of asyncio.gather() so,
+        # if any task raises an unexpected exception,
+        # it cancels every other task and raises an exception to signal the bug.
+        for protocol_id in expected_protocol_ids:
+            protocol_subdirectory = protocols_directory / protocol_id
+            task_group.start_soon(compute_source, protocol_id, protocol_subdirectory)
 
-def _convert_sql_row_to_resource(sql_row: sqlalchemy.engine.Row) -> ProtocolResource:
+    for id in expected_protocol_ids:
+        assert id in sources_by_id
```

Comment:
Take this or leave it, I'm just trying to figure out which I prefer 🙃 

```suggestion
    assert set(sources_by_id.keys()) == expected_protocol_ids
```